machineconfiguration/v1/types.go

```diff
@@ -759,3 +760,123 @@ type ContainerRuntimeConfigList struct {
 
 	Items []ContainerRuntimeConfig `json:"items"`
 }
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineConfigState describes the health of the Machines on the system
+//
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=1
+type MachineConfigState struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +kubebuilder:validation:Required
+	Spec MachineConfigStateSpec `json:"spec"`
+	// +optional
+	Status MachineConfigStateStatus `json:"status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineConfigStateList describes all of the MachinesStates on the system
+//
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=1
+type MachineConfigStateList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []MachineConfigState `json:"items"`
+}
+
+// MachineConfigStateSpec describes the type of State we are managing
+type MachineConfigStateSpec struct {
```

Comment:
Given that we are now making the MachineConfigState specific to a state. What fields would we want to see in the spec? IMO we probably want to see information in the spec about the node, so a `nodeRef`, and then probably information about the pool and the desired configuration maybe?

What information does the Node currently get populated with by MCO bits today?

Reply:
The node only gets populated with MCO annotations about the desired/current MachineConfig, a reason/state annotation that can be filled with anything to help us parse what is going on in the MSC, and then drainer information. 

So yeah, I can add some of that stuff here so people can view it in the "state reporting object". NodeRef make sense. For desired configuration do you mean desired MachineConfig or desired config of this MachineState? I can also attach which pool this is related to, but that will live in the name (upgrade-worker vs upgrade-master)

Reply:
would it be alright to do an array of NodeRefs? since each MachineConfigState is related to a pool not a specific node we need at least 3 nodeRefs per MachineConfigState usually.

Reply:
Discussed in slack, we have an example of a single 1:1 mapping of node to MSC and that seems to be preferable